packages/protocol/contracts/L2/TaikoL2.sol

```diff
@@ -30,50 +42,95 @@ contract TaikoL2 is AddressResolver, ReentrancyGuard, IHeaderSync {
+    uint128 public yscale;
+    uint64 public xscale;
+    uint64 public gasIssuedPerSecond;
 
-    uint256[46] private __gap;
+    uint64 public parentTimestamp;
+    uint64 public latestSyncedL1Height;
+    uint64 public gasExcess;
+    uint64 public __reserved1;
+
+    uint256[45] private __gap;
 
     /**********************
      * Events and Errors  *
      **********************/
 
-    event BlockInvalidated(bytes32 indexed txListHash);
-
-    error L2_INVALID_SENDER();
+    // Captures all block variables mentioned in
+    // https://docs.soliditylang.org/en/v0.8.18/units-and-global-variables.html
+    event BlockVars(
+        uint64 number,
+        uint64 basefee,
+        uint64 gaslimit,
+        uint64 timestamp,
+        bytes32 parentHash,
+        uint256 prevrandao,
+        address coinbase,
+        uint32 chainid
+    );
+
+    error L2_BASEFEE_MISMATCH(uint64 expected, uint64 actual);
+    error L2_INVALID_1559_PARAMS();
     error L2_INVALID_CHAIN_ID();
-    error L2_INVALID_GAS_PRICE();
-    error L2_PUBLIC_INPUT_HASH_MISMATCH();
+    error L2_INVALID_SENDER();
+    error L2_PUBLIC_INPUT_HASH_MISMATCH(bytes32 expected, bytes32 actual);
+    error L2_TOO_LATE();
+
+    error M1559_UNEXPECTED_CHANGE(uint64 expected, uint64 actual);
+    error M1559_OUT_OF_STOCK();
 
     /**********************
      * Constructor         *
      **********************/
 
-    constructor(address _addressManager) {
-        if (block.chainid == 0) revert L2_INVALID_CHAIN_ID();
-        AddressResolver._init(_addressManager);
-
-        bytes32[255] memory ancestors;
-        uint256 number = block.number;
-        for (uint256 i; i < 255 && number >= i + 2; ++i) {
-            ancestors[i] = blockhash(number - i - 2);
+    function init(
+        address _addressManager,
+        EIP1559Params calldata _param1559
+    ) external initializer {
+        if (block.chainid <= 1 || block.chainid >= type(uint32).max)
+            revert L2_INVALID_CHAIN_ID();
+        if (block.number > 1) revert L2_TOO_LATE();
+
+        if (_param1559.gasIssuedPerSecond != 0) {
+            if (
+                _param1559.basefee == 0 ||
+                _param1559.gasExcessMax == 0 ||
+                _param1559.gasTarget == 0 ||
+                _param1559.ratio2x1x == 0
+            ) revert L2_INVALID_1559_PARAMS();
+
+            uint128 _xscale;
+            (_xscale, yscale) = Lib1559Math.calculateScales({
+                xExcessMax: _param1559.gasExcessMax,
+                price: _param1559.basefee,
+                target: _param1559.gasTarget,
+                ratio2x1x: _param1559.ratio2x1x
+            });
+
+            if (_xscale == 0 || _xscale >= type(uint64).max || yscale == 0)
+                revert L2_INVALID_1559_PARAMS();
+            xscale = uint64(_xscale);
+
+            // basefee = _param1559.basefee;
+            gasIssuedPerSecond = _param1559.gasIssuedPerSecond;
+            gasExcess = _param1559.gasExcessMax / 2;
```

Comment:
Not sure how big `gasExcessMax`, but seems like it should be a pretty reasonable small value otherwise `gasExcess` will produce a very high basefee that will only go down very slowly (potentially never). Maybe just using `0` here safer?

Reply:
I think maybe these two options are mathematically identical, one is to set the initial basefee to a best-guess value, the other chooses 0 as the value then wait it to go up gradually.

Reply:
After the initial deployment, if not a lot of gas is used, then the price will gradually decrease to zero, your understanding is correct.